tests/integration/test_backup_restore_on_cluster/test_cancel_backup.py

```diff
@@ -0,0 +1,614 @@
+        )
+    else:
+        nodes_to_consider = [node] if node is not None else nodes
+        for n in nodes_to_consider:
+            count = get_num_system_processes(
+                node=n, backup_id=backup_id, restore_id=restore_id
+            )
+            print(
+                f"{get_node_name(n)}: Got {count} system processes for {operation_name} {id} after {waited} seconds"
+            )
+        assert False
+    return waited
+
+
+# Kills a BACKUP or RESTORE query.
+# Returns how many seconds the KILL QUERY was executing.
+def kill_query(node, backup_id=None, restore_id=None, is_initial_query=None, timeout=2):
+    id = backup_id if backup_id is not None else restore_id
+    query_kind = "Backup" if backup_id is not None else "Restore"
+    operation_name = "backup" if backup_id is not None else "restore"
+    print(f"{get_node_name(node)}: Cancelling {operation_name} {id}")
+    filter_for_is_initial_query = (
+        f" AND (is_initial_query = {is_initial_query})"
+        if is_initial_query is not None
+        else ""
+    )
+    node.query(
+        f"KILL QUERY WHERE (query_kind='{query_kind}') AND (query LIKE '%{id}%'){filter_for_is_initial_query} SYNC"
+    )
+    node.query("SYSTEM FLUSH LOGS")
+    duration = (
+        int(
+            node.query(
+                f"SELECT query_duration_ms FROM system.query_log WHERE query_kind='KillQuery' AND query LIKE '%{id}%' AND type='QueryFinish'"
+            )
+        )
+        / 1000
+    )
+    print(
+        f"{get_node_name(node)}: Cancelled {operation_name} {id} after {duration} seconds"
+    )
+    return duration
+
+
+# Stops all ZooKeeper servers.
+def stop_zookeeper_servers():
+    print("Stopping all ZooKeeper servers")
+    cluster.stop_zookeeper_nodes(["zoo1", "zoo2", "zoo3"])
+
+
+# Starts all ZooKeeper servers back.
+def start_zookeeper_servers():
+    print("Starting all ZooKeeper servers")
+    cluster.start_zookeeper_nodes(["zoo1", "zoo2", "zoo3"])
+
+
+# Disconnects a specified host with ZooKeeper.
+def disable_zookeeper_on_host(node):
+    print(f"Disconnecting {get_node_name(node)} with ZooKeeper")
+    node.rename_config(
+        "/etc/clickhouse-server/conf.d/zookeeper_config.xml",
+        "zookeeper_config.xml.disabled",
+    )
+    node.query("SYSTEM RELOAD CONFIG")
+
+
+# Connects a specified node with ZooKeeper back.
+def enable_zookeeper_on_host(node):
+    print(f"Connecting {get_node_name(node)} with ZooKeeper back")
+    node.rename_config(
+        "/etc/clickhouse-server/conf.d/zookeeper_config.xml.disabled",
+        "zookeeper_config.xml",
+    )
+    node.query("SYSTEM RELOAD CONFIG")
+
+
+# Sleeps for random amount of time.
+def random_sleep(max_seconds):
+    sleep_time = random.uniform(0, max_seconds)
+    print(f"Sleeping {sleep_time} seconds")
+    time.sleep(sleep_time)
+
+
+# Checks that BACKUP and RESTORE cleaned up properly with no trash left in ZooKeeper, backups folder, and logs.
```

Comment:
These tests also check that a failed or cancelled backup or restore cleans up correctly leaving no trash in ZooKeeper or in the destination storage.